File: ios/App/App/GameAudioPlugin.swift
import Capacitor

@objc(GameAudioPlugin)
public class GameAudioPlugin: CAPPlugin {
    @objc func start(_ call: CAPPluginCall) {
        if let volume = call.getFloat("bgmVolume") {
            GameAudioManager.shared.setBgmVolume(volume)
        }
        GameAudioManager.shared.startBgm { started in
            call.resolve(["started": started])
        }
    }

    @objc func pauseBgm(_ call: CAPPluginCall) {
        GameAudioManager.shared.pauseBgm()
        call.resolve()
    }

    @objc func prepareForRecording(_ call: CAPPluginCall) {
        let resumeBgm = call.getBool("resumeBgm") ?? true
        let keepBgm = call.getBool("keepBgm") ?? false
        let webManaged = call.getBool("webManaged") ?? false
        GameAudioManager.shared.prepareForRecording(resumeBgm: resumeBgm, keepBgm: keepBgm, webManaged: webManaged)
        call.resolve()
    }

    @objc func finishRecording(_ call: CAPPluginCall) {
        let resumeBgm = call.getBool("resumeBgm") ?? true
        let keepBgm = call.getBool("keepBgm") ?? false
        let webManaged = call.getBool("webManaged") ?? false
        GameAudioManager.shared.finishRecording(resumeBgm: resumeBgm, keepBgm: keepBgm, webManaged: webManaged)
        call.resolve()
    }

    @objc func setBgmVolume(_ call: CAPPluginCall) {
        let volume = call.getFloat("volume") ?? 0.3
        GameAudioManager.shared.setBgmVolume(volume)
        call.resolve()
    }

    @objc func playSfx(_ call: CAPPluginCall) {
        guard let assetId = call.getString("assetId"), let fileName = call.getString("fileName") else {
            call.reject("Missing assetId or fileName")
            return
        }

        let volume = call.getFloat("volume") ?? 1
        GameAudioManager.shared.playSfx(assetId: assetId, fileName: fileName, volume: volume)
        call.resolve()
    }

    @objc func playBundleAudio(_ call: CAPPluginCall) {
        guard let relativePath = call.getString("relativePath") else {
            call.reject("Missing relativePath")
            return
        }

        let volume = call.getFloat("volume") ?? 1
        GameAudioManager.shared.playBundleAudio(relativePath: relativePath, volume: volume) { played in
            call.resolve(["played": played])
        }
    }

    @objc func playRemoteAudio(_ call: CAPPluginCall) {
        guard let url = call.getString("url") else {
            call.reject("Missing url")
            return
        }

        let volume = call.getFloat("volume") ?? 1
        GameAudioManager.shared.playRemoteAudio(urlString: url, volume: volume) { played in
            call.resolve(["played": played])
        }
    }

    @objc func preloadRemoteAudio(_ call: CAPPluginCall) {
        guard let url = call.getString("url") else {
            call.reject("Missing url")
            return
        }

        GameAudioManager.shared.preloadRemoteAudio(urlString: url) { preloaded in
            call.resolve(["preloaded": preloaded])
        }
    }

    @objc func stopBundleAudio(_ call: CAPPluginCall) {
        GameAudioManager.shared.stopBundleAudio()
        call.resolve()
    }

    @objc func stopRemoteAudio(_ call: CAPPluginCall) {
        GameAudioManager.shared.stopRemoteAudio()
        call.resolve()
    }

    @objc func startSpeechCapture(_ call: CAPPluginCall) {
        let maxDurationMs = call.getInt("maxDurationMs") ?? 10000
        let noVoiceTimeoutMs = call.getInt("noVoiceTimeoutMs") ?? 5000
        let minVoiceWindowMs = call.getInt("minVoiceWindowMs") ?? 2200
        let silenceMs = call.getInt("silenceMs") ?? 1000
        let tailBufferMs = call.getInt("tailBufferMs") ?? 450
        let resumeBgm = call.getBool("resumeBgm") ?? true

        GameAudioManager.shared.startSpeechCapture(
            maxDurationMs: maxDurationMs,
            noVoiceTimeoutMs: noVoiceTimeoutMs,
            minVoiceWindowMs: minVoiceWindowMs,
            silenceMs: silenceMs,
            tailBufferMs: tailBufferMs,
            resumeBgm: resumeBgm,
            onLevel: { [weak self] payload in
                self?.notifyListeners("speechCaptureLevel", data: payload)
            }
        ) { result in
            switch result {
            case .success(let payload):
                call.resolve(payload)
            case .failure(let error):
                call.reject(error.localizedDescription, nil, error)
            }
        }
    }

    @objc func stopSpeechCapture(_ call: CAPPluginCall) {
        let reason = call.getString("reason") ?? "manual"
        GameAudioManager.shared.stopSpeechCapture(reason: reason)
        call.resolve(["stopping": true])
    }

    @objc func preloadSfx(_ call: CAPPluginCall) {
        guard let assetId = call.getString("assetId"), let fileName = call.getString("fileName") else {
            call.reject("Missing assetId or fileName")
            return
        }

        let volume = call.getFloat("volume") ?? 1
        GameAudioManager.shared.preloadSfx(assetId: assetId, fileName: fileName, volume: volume) { preloaded in
            call.resolve(["preloaded": preloaded])
        }
    }
}

File: android/app/src/main/java/com/enguistics/vocabconqueror/GameAudioPlugin.java
package com.enguistics.vocabconqueror;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "GameAudio")
public class GameAudioPlugin extends Plugin {
    @PluginMethod
    public void start(PluginCall call) {
        double volume = call.getDouble("bgmVolume", 0.3);
        boolean started = GameAudioManager.getInstance().startBgm(getContext(), (float) volume);
        call.resolve(new JSObject().put("started", started));
    }

    @PluginMethod
    public void pauseBgm(PluginCall call) {
        GameAudioManager.getInstance().pauseBgm();
        call.resolve();
    }

    @PluginMethod
    public void prepareForRecording(PluginCall call) {
        boolean resumeBgm = call.getBoolean("resumeBgm", true);
        GameAudioManager.getInstance().prepareForRecording(resumeBgm);
        call.resolve();
    }

    @PluginMethod
    public void finishRecording(PluginCall call) {
        boolean resumeBgm = call.getBoolean("resumeBgm", true);
        GameAudioManager.getInstance().finishRecording(getContext(), resumeBgm);
        call.resolve();
    }

    @PluginMethod
    public void setBgmVolume(PluginCall call) {
        double volume = call.getDouble("volume", 0.3);
        GameAudioManager.getInstance().setBgmVolume((float) volume);
        call.resolve();
    }

    @PluginMethod
    public void playBundleAudio(PluginCall call) {
        String relativePath = call.getString("relativePath", "");
        double volume = call.getDouble("volume", 1.0);
        if (relativePath.isEmpty()) {
            call.reject("Missing relativePath");
            return;
        }

        GameAudioManager.getInstance().playBundleAudio(getContext(), relativePath, (float) volume, played -> {
            JSObject result = new JSObject();
            result.put("played", played);
            call.resolve(result);
        });
    }

    @PluginMethod
    public void playRemoteAudio(PluginCall call) {
        String url = call.getString("url", "");
        double volume = call.getDouble("volume", 1.0);
        if (url.isEmpty()) {
            call.reject("Missing url");
            return;
        }

        GameAudioManager.getInstance().playRemoteAudio(getContext(), url, (float) volume, played -> {
            JSObject result = new JSObject();
            result.put("played", played);
            call.resolve(result);
        });
    }

    @PluginMethod
    public void preloadRemoteAudio(PluginCall call) {
        String url = call.getString("url", "");
        if (url.isEmpty()) {
            call.reject("Missing url");
            return;
        }

        GameAudioManager.getInstance().preloadRemoteAudio(getContext(), url, preloaded -> {
            JSObject result = new JSObject();
            result.put("preloaded", preloaded);
            call.resolve(result);
        });
    }

    @PluginMethod
    public void stopBundleAudio(PluginCall call) {
        GameAudioManager.getInstance().stopBundleAudio();
        call.resolve();
    }

    @PluginMethod
    public void stopRemoteAudio(PluginCall call) {
        GameAudioManager.getInstance().stopRemoteAudio();
        call.resolve();
    }

    @PluginMethod
    public void playSfx(PluginCall call) {
        String fileName = call.getString("fileName", "");
        double volume = call.getDouble("volume", 1.0);
        if (fileName.isEmpty()) {
            call.reject("Missing fileName");
            return;
        }

        GameAudioManager.getInstance().playSfx(getContext(), fileName, (float) volume);
        call.resolve(new JSObject().put("played", true));
    }

    @PluginMethod
    public void preloadSfx(PluginCall call) {
        String fileName = call.getString("fileName", "");
        if (fileName.isEmpty()) {
            call.reject("Missing fileName");
            return;
        }

        GameAudioManager.getInstance().preloadSfx(getContext(), fileName, preloaded -> {
            JSObject result = new JSObject();
            result.put("preloaded", preloaded);
            call.resolve(result);
        });
    }
}
